data/transactions/logic/opcodes.go

```diff
@@ -488,34 +488,34 @@ func (spec *OpSpec) deadens() bool {
 // assembly-time, with ops.returns()
 var OpSpecs = []OpSpec{
 	{0x00, "err", opErr, proto(":x"), 1, detDefault()},
-	{0x01, "sha256", opSHA256, proto("b:3"), 1, costly(7)},
-	{0x02, "keccak256", opKeccak256, proto("b:3"), 1, costly(26)},
-	{0x03, "sha512_256", opSHA512_256, proto("b:3"), 1, costly(9)},
+	{0x01, "sha256", opSHA256, proto("b:b{32}"), 1, costly(7)},
+	{0x02, "keccak256", opKeccak256, proto("b:b{32}"), 1, costly(26)},
+	{0x03, "sha512_256", opSHA512_256, proto("b:b{32}"), 1, costly(9)},
 
 	// Cost of these opcodes increases in AVM version 2 based on measured
 	// performance. Should be able to run max hashes during stateful TEAL
 	// and achieve reasonable TPS. Same opcode for different versions
 	// is OK.
-	{0x01, "sha256", opSHA256, proto("b:3"), 2, costly(35)},
-	{0x02, "keccak256", opKeccak256, proto("b:3"), 2, costly(130)},
-	{0x03, "sha512_256", opSHA512_256, proto("b:3"), 2, costly(45)},
+	{0x01, "sha256", opSHA256, proto("b:b{32}"), 2, costly(35)},
+	{0x02, "keccak256", opKeccak256, proto("b:b{32}"), 2, costly(130)},
+	{0x03, "sha512_256", opSHA512_256, proto("b:b{32}"), 2, costly(45)},
 
 	/*
 		Tabling these changes until we offer unlimited global storage as there
 		is currently a useful pattern that requires hashes on long slices to
 		creating logicsigs in apps.
 
-		{0x01, "sha256", opSHA256, proto("b:b"), ?, costByLength(...)},
-		{0x02, "keccak256", opKeccak256, proto("b:b"), ?, costByLength(...)},
-		{0x03, "sha512_256", opSHA512_256, proto("b:b"), ?, costByLength(...)},
+		{0x01, "sha256", opSHA256, proto("b:b{32}"), ?, costByLength(...)},
+		{0x02, "keccak256", opKeccak256, proto("b:b{32}"), ?, costByLength(...)},
+		{0x03, "sha512_256", opSHA512_256, proto("b:b{32}"), ?, costByLength(...)},
 	*/
 
-	{0x04, "ed25519verify", opEd25519Verify, proto("b63:T"), 1, costly(1900).only(ModeSig)},
-	{0x04, "ed25519verify", opEd25519Verify, proto("b63:T"), 5, costly(1900)},
+	{0x04, "ed25519verify", opEd25519Verify, proto("bb{64}b{32}:T"), 1, costly(1900).only(ModeSig)},
+	{0x04, "ed25519verify", opEd25519Verify, proto("bb{64}b{32}:T"), 5, costly(1900)},
 
-	{0x05, "ecdsa_verify", opEcdsaVerify, proto("3bbbb:T"), 5, costByField("v", &EcdsaCurves, ecdsaVerifyCosts)},
-	{0x06, "ecdsa_pk_decompress", opEcdsaPkDecompress, proto("b:bb"), 5, costByField("v", &EcdsaCurves, ecdsaDecompressCosts)},
-	{0x07, "ecdsa_pk_recover", opEcdsaPkRecover, proto("3i33:bb"), 5, field("v", &EcdsaCurves).costs(2000)},
+	{0x05, "ecdsa_verify", opEcdsaVerify, proto("b{32}b{32}b{32}b{32}b{32}:T"), 5, costByField("v", &EcdsaCurves, ecdsaVerifyCosts)},
```

Comment:
I understand the message size of 32 and signature (r,s) but looks like there is no such requirement for x,y

Reply:
It's true that our `ecdsa_verify` does not require 32 byte values.  We offer pretty low-level interface where we interpret `x` and `y` as big-endian integers, and it is conceivable that your `x` or `y` only needs 31 bytes to be encoded properly.  But, if we think about how you're actually going get your `x` or `y`,  you'll pretty much always have a 32 byte sequence anyway, even if they would fit in 31 or less.  I think you will get your values from one of two places.

* You use `ecdsa_pk_decompress` on a 33 byte value, and it spits out `x` and `y`.  In such cases, we _always_ return 32 byte values, even if a smaller encoding were possible. We use `leadingZeros(32, x)`

* You accept some external encoding for the uncompressed `x`, `y` pair.  In the encodings I've seen, they are encoded as fixed sized, 32 byte sequences.  But I admit that, in principle, they could be dynamically sized.

I'm willing to remove the restriction, since it's not literally required, but I think it's better to have it, as it helps guide the user to use `ecdsa_pk_decompress` based on matching types.

Reply:
Final thoughts? Think it should stay or go?